Add additional definitions of 'string' (w3c/i18n-actions#41)

aphillips opened the pull request:
The working group discussed this in teleconference on 2023-09-21 and decided to rewrite this section to address our concerns.

---

There are multiple sources for the term 'string', including WEBIDL,
INFRA, Unicode, and XML Schema. In our discussions with RDF-star,
we have found that there is a need for generic string definitions
for use by JSON, Schema.org, RDF, and XML-based specs which are not
consistent with DOMString/USVString.

This PR needs to be accompanied by changes to the I18N-GLOSSARY in order
to be complete. This PR is for initial discussion purposes before we
embark on those changes.

Mentioning w3ctag/design-principles#454 because that issue tracks our coordination with TAG


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/aphillips/bp-i18n-specdev/pull/117.html" title="Last updated on Oct 19, 2023, 3:14 PM UTC (6563800)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/bp-i18n-specdev/117/2f5c739...aphillips:6563800.html" title="Last updated on Oct 19, 2023, 3:14 PM UTC (6563800)">Diff</a>

aphillips committed:
Add additional definitions for 'string' (w3c/i18n-actions#41)

There are multiple sources for the term 'string', including WEBIDL,
INFRA, Unicode, and XML Schema. In our discussions with RDF-star,
we have found that there is a need for generic string definitions
for use by JSON, Schema.org, RDF, and XML-based specs which are not
consistent with DOMString/USVString.

This PR needs to be accompanied by changes to the I18N-GLOSSARY in order
to be complete. This PR is for initial discussion purposes before we
embark on those changes.

aphillips committed:
Merge branch 'w3c:gh-pages' into gh-pages

aphillips committed:
Wording improvements

aphillips committed:
Merge remote-tracking branch 'refs/remotes/origin/gh-pages' into gh-pages

aphillips committed:
First pass at telecon changes

aphillips committed:
Further edits

aphillips committed:
Further improvements.

aphillips committed:
Make the DOMString vs USVString paragraph less repetitive

xfq commented on index.html:
```diff
@@ -1909,36 +1909,53 @@ <h3>Defining 'string'</h3>
+
+<p>The best practices found in this section are mutually consistent with those in [[DESIGN-PRINCIPLES]]. The definitions in this section use terms found in the <cite>Internationalization Glossary</cite> [[I18N-GLOSSARY]]. Some of these definitions are taken directly from [[WEBIDL]], [[INFRA]], or the Unicode glossary; in which case the definitions are quoted verbatim and include links to their so ...
+
+<p>[[UNICODE]] forms the basis for text on the Web, so the basic definition of a string is: <em>a well-formed sequence of zero or more <a>Unicode Scalar Values</a></em>. Because the term 'string' is ambiguous, this definition can be more specifically referred to using one of these definitions:
```
And CSS specifications can refer to the definition in https://www.w3.org/TR/css-values-4/#strings ?

xfq replied:
And some specs refer to RFC 3629 directly, not sure if that's something we want to recommend, though: https://w3c.github.io/IFT/Overview.html#primitives

aphillips replied:
3629 is the (IETF) definition of UTF-8, not string.

CSS's definition is weird, since it appears to be a USVString but points to Infra's (DOMString) definition of string. 

I don't think our set of references needs to be exhaustive and fewer references is actually better for our customers.

xfq commented on index.html:
```diff
@@ -1909,36 +1909,53 @@ <h3>Defining 'string'</h3>
+
+<p>The [[DOM]], JavaScript (including its APIs and document formats), or certain platforms use the <a>UTF-16</a> <a>character encoding</a> to store and process string data. This can expose the underlying details of the UTF-16 encoding, leading to the need to use a different type of string, the <a>DOMString</a> [[WEBIDL]]. For example, [[INFRA]] defines a string this way:</p>
+
+<p style="background-color:white; border: 1px solid brown; margin:0.5em">A string is a sequence of unsigned 16-bit integers, also known as <a>code units</a></p>
```
Maybe add a `padding: 0.5em;` to this box?

aphillips replied:
Done.

aphillips committed:
Address comments

aphillips committed:
Merge branch 'w3c:gh-pages' into gh-pages

aphillips committed:
Further (deep) rewrite

This change significantly reorganizes the section on strings in hopes
of making the definitions clearer and to reduce the overall
repetitive nature of the text.

It also moves the text about design-principles to a note at the top,
rather than burying it in the text.

The relationship of DOMString and USVString to surrogates is clarified
and pulled into one place.

aphillips committed:
Further improvements

- Move inline style to local.css
- fix lint-ignore on string definition
- link to webidl definitions as needed
- add a note about why you want DOMString when specifying bytes on the
  wire

aphillips committed:
Actually link WebIDL's definitions

himorin commented on index.html:
```diff
@@ -1915,47 +1915,87 @@ <h3>Defining 'string'</h3>
+
 <div class="req" id="char_string_domstring">
-	<p class="advisement">When designing a web platform feature which operates on strings, use <a href="https://webidl.spec.whatwg.org/#idl-DOMString">DOMString</a> unless you have a specific reason not to.</p>
+	<p class="advisement">When designing a Web platform feature which operates on strings, use <a href="https://webidl.spec.whatwg.org/#idl-DOMString">DOMString</a> unless you have a specific reason not to.</p>
```
I suppose we can replace `<a>` with `{{DOMString}}`?

aphillips replied:
Done.

himorin commented on index.html:
```diff
@@ -1915,47 +1915,87 @@ <h3>Defining 'string'</h3>
+
+<p>By contrast, a {{DOMString}} allows unpaired <a>surrogate</a> <a>code units</a> to appear in a string. An unpaired surrogate can result in processing errors or it must be replaced with the Unicode replacment character (<span class="codepoint" translate="no"><bdi lang="und">&#xFFFD;</bdi><code class="uname">U+FFFD REPLACEMENT CHARACTER</code></span>) when processing the characters inside the st ...
+
+<p>In well-formed <a>UTF-8</a>, <a>surrogate</a> <a>code points</a> should not be encoded and isolated surrogates are replaced by the replacement character. <a>Legacy character encodings</a> never encode <a>surrogates</a>.</p>
```
I'm not sure whether `should not be encoded` is correct per bit handling point of view... I believe mapping to UTF-8 is from U+10XXXX but not surrogate coded UTF-16 form... (so, when I wrote some code, first moving to replacement character and 32bit form, and go for UTF-8 bit operation...)

aphillips replied:
Yes, that's right. In well-formed UTF-8, surrogate _code points_ should be encoded as U+FFFD. (A UTF-16 surrogate pair would be encoded as a four-byte UTF-8 sequence). There might be coders that produce an isolated 3-byte surrogate, I suppose? Let me try to make this clearer.

himorin replied:
thank you for checking!

aphillips committed:
Address @himoren's comments

- Fix links to DOMString and USVString to use WebIDL markup
- Rephrase the sentence about UTF-8's relationship to surrogate code
  points.

aphillips committed:
Further rewriting of string section per 2023-10-05 call

aphillips committed:
Move the mustard ahead of the explanation

We have a general practice of putting the best practice text
before the explanation of the BP so that users linking into
specdev can read below the best practice rather than having
to scroll up and start the section over.

This change moves the string definition guidance ahead of all
of the text about string types.

aphillips committed:
Add missing close parenthesis

aphillips committed:
Address comments in 2023-10-12 call

- Change the MUSTard to recommend USVString first.
- Split the MUSTard into two BPs
- Add a missing link around one instance of `USVString`
- Add an issue that includes a link to w3ctag/design-principles#454 and
  appropriately scary sounding text.

aphillips committed:
Improve the DOMString requirement

r12a commented on index.html:
```diff
@@ -1928,7 +1928,7 @@ <h3>Defining 'string'</h3>
 <div class="req" id="char_string_dom">
-	<p class="advisement">If your specification pertains to the [[DOM]], [[HTML]], or uses [[WebIDL]], or defines strings as opaque values that are not processed, use {{DOMString}} as a definition of string.</p>
+	<p class="advisement">Use a string definition consistent with {{DOMString}} if your specification does not process the internal value of strings and is not required to check for unpaired surrogate code points, or if your specification pertains to the [[DOM]], defines a JavaScript API or data format, or defines strings as opaque values that are not processed.</p>
 	<details class="links"><summary>explantations &amp; examples</summary>
```
i could be wrong, but **explantations** probably needs to be changed to **explanations**

r12a replied:
otherwise, i think this is probably ok, thanks

aphillips replied:
🤦

Good catch. Sadly, I *copied* that text from elsewhere in the previous doc to avoid making typos. Or is an _explantation_ an explanation from the plantation?

aphillips committed:
Fix typo

aphillips committed:
Make change proposed in 2023-10-19 teleconference

aphillips committed:
Merge branch 'gh-pages' into gh-pages
